### DISCUSSION TITLE
FileUpload to s3 issue

### DISCUSSION THREAD

**Discussion by liza-nt** [filamentphp/filament#8361] [Help]
I am uploading image to s3 and i can't able find out the reason.. it is not even giving an error in the network..so is there anywhere i can print or return a message while uploading if my upload is successful or not

**Comment by leandrocfe** [comment]
Did you follow these steps? [configuring-the-storage-disk-and-directory](https://filamentphp.com/docs/3.x/forms/fields/file-upload#configuring-the-storage-disk-and-directory)
Can you share more details about the issue?

**Comment by liza-nt** [comment]
```
 FileUpload::make('kheltag_profile_image')
                                    ->label("Upload Profile Image")
                                    ->disk('s3')
                                    ->directory('uploads/coach/profile_pic/large')
                                    ->visibility('private')
                                    ->preserveFilenames()
                                    ->image()
                                    ```
This is my code...when I am uploading image I dont know its being uploaded or not its not giving any error in network also

**Reply by leandrocfe** [reply]
What happens if you remove `visibility('private')`?

**Reply by liza-nt** [reply]
in local i have my own credentials of s3 bucket ..with visibility('private') its not uploading...in production I do not have access of the bucket but only have credentials so i do not know it will work or not without it

**Reply by leandrocfe** [reply]
I think you could try [minio](https://dev.to/gjrdiesel/set-up-minio-w-laravel-local-s3-2cj4) in your local. I'm using with [laravel sail](https://laravel.com/docs/10.x/sail) and it is working
```bash
FILESYSTEM_DRIVER=s3
AWS_ACCESS_KEY_ID=sail
AWS_SECRET_ACCESS_KEY=password
AWS_DEFAULT_REGION=us-east-1
AWS_BUCKET=local
AWS_ENDPOINT=http://minio:9000
AWS_URL=http://localhost:9000/local
AWS_USE_PATH_STYLE_ENDPOINT=true
```

**Comment by ssatz** [comment]
Even i am having the same issue. But it is working if i use `Storage::disk('s3')->put('/',$content)`

**Comment by jgmuchiri** [comment]
Had the same issue turned out I had to recompile assets `npm run build` and update bucket CORS with the following

```json
[
    {
        "AllowedHeaders": [
            "*"
        ],
        "AllowedMethods": [
            "GET",
            "HEAD"
        ],
        "AllowedOrigins": [
            "*"
        ],
        "ExposeHeaders": [],
        "MaxAgeSeconds": 3000
    }
]
```

I also updated Object Ownership with "ACL Enabled" -> "Object Writer"

**Reply by jerexbambex** [reply]
Thank you this worked for me.

**Comment by brentkelly** [comment]
I'm not 100% sure this is the same issue, but I will leave this here in case others are running into an issue where S3 Uploads are simply not working using the default `FileUpload` component.

By default livewire attempts to use S3 ACL if you set to public `visibility`. This will throw an error unless your S3 bucket is configured to use ACLs - something that S3 recommends against as the default. The trouble is - by default laravel is configured to suppress errors for s3 - so your upload will silently fail and you will be left scratching your head mouthing a quiet WT... 😕🤷. (To change that, set `throw` to `true` in your `config/filesystems.php`).

If you set visiblity to `private`, livewire will use temporaryUrls for reading images which isn't really what we're after either. So ... checkmate.

The good news is there is a bit of a dirty hack workaound: setting this to neither public nor private allows you to treat uploads as public, but without trying to force file-specific ACL - instead relying on your bucket policy (to allow `s3:GetObject`). 

This is pretty much what I'm always after so in my Filament projects that use S3 I create my own `FileUpload` component that inherits from the base filament one:

```php
use Closure;
use Filament\Forms\Components\FileUpload as BaseFileUpload;

class FileUpload extends BaseFileUpload
{
    /**
     * By default livewire attempts to use S3 ACL if you set to public visibility. This
     * will throw an error unless your S3 bucket is configured to use ACLs - something
     * that S3 recommends against as the default. If you set visiblity to private,
     * livewire will use temporaryUrls for reading images. Setting this to neither
     * public or private allows you to treat uploads as public, but without trying
     * to force file-specific ACL - instead relying on your bucket policy (to allow
     * s3:GetObject)
     */
    protected string|Closure $visibility = 'bucket-defaults';
}
```

`bucket-defaults` is just any random string. It can even be an empty string - so long as it isn't `public` or `private`.

If I use that component in place of the Filament `FileUpload`, everything now works as I would expect.
